src/util/string.h

```diff
@@ -17,6 +17,55 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <unsigned num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { Detail_CheckNumFormatSpecifiers(fmt); }
+    constexpr static void Detail_CheckNumFormatSpecifiers(std::string_view str)
+    {
+        unsigned count_normal{0}; // Number of "normal" specifiers, like %s
+        unsigned count_pos{0};    // Max number in positional specifier, like %8$s
+        for (auto it{str.begin()}; it < str.end();) {
+            if (*it != '%') {
+                ++it;
+                continue;
+            }
+
+            if (++it >= str.end()) throw "Format specifier incorrectly terminated by end of string";
+            if (*it == '%') {
+                // %%
+                ++it;
+                continue;
+            }
+
+            // Test for positional specifier, like %8$
```

Comment:
Absolutely, but we may not want to spend time implementing positional parsing, if we agree that we don't want to encourage their usage

Reply:
I think positional args are more readable than repeated args (even in this example which is atypically long) because:

- People have limited working memories. it is a easier to see 4 parameters passed and remember what numbers 1-4 are than to see 6 parameters passed and remember what 1-6 are. Remembering a list is also harder if the list has duplicate values.
- Positional arguments are more explicit. "Configuration file %3$s" it is clearer than "Configuration file %s" because it tells you that the 3rd argument is substituted. You don't have to manually scan the string and count earlier %s.
- Positional arguments make it easier understand messages and see whether they are internally consistent. "Data directory %1$s contains a %2$s file [...] Delete or rename the %2$s file in data directory %1$s" is easier to understand and verify than the alternative with %s everywhere.

Reply:
Thanks @ryanofsky, it seems there is a need for that after all, please resolve this comment.